### DISCUSSION TITLE
Requesting help on setting up RavenDB with Spring boot application

### DISCUSSION THREAD

**Discussion by rahul-raj** [ravendb/ravendb#18639] [General]
Hello All,

I'm pretty much new to RavenDB and currently going through all the documentation to setup a sample document collection.
I have prior experience to MongoDB and trying to relate the same here as well. 
I have below questions.

1) In RavenDB studio, I see that we can create new database and can add documents manually, which is great. But, I don't see an option there to create collections. It creates a collection `@empty` by default and all documents goes underneath it. 
I understand that we may be able to create collections using Java code. But I wanted to know how to create from the studio. 
I want to create collections from studio and then add/view documents from the application(Java client). 

![Check this screenshot](https://github.com/ravendb/ravendb/assets/517415/ed780d69-6f18-427a-a4ba-3160ff6eece9)

2) How do I set a collection name using Java code ? 

3)   I tried to create a singleton session like this

```
@Bean
IDocumentSession getSession() throws KeyStoreException, IOException, CertificateException, NoSuchAlgorithmException {
		KeyStore clientStore = KeyStore.getInstance("PKCS12");
		clientStore.load(new FileInputStream("C:\\certificate.with.password.pfx"), "D146AB87B64E3D2A23D3AD6606058A".toCharArray());
		try (DocumentStore store = new DocumentStore())  {
			store.setCertificate(clientStore);
			store.setDatabase("test");
			store.setUrls(new String[]{ "https://mydomain.ravendb.cloud" });
			store.initialize();
			return  store.openSession();
		}

	}
```

And then tried to insert a document using a POST endpoint:

```
@RestController
public class TestController {
    @Autowired
    IDocumentSession session;

    @PostMapping("/test")
    public Job test(@RequestBody Job job) {
        session.store(job);
        session.saveChanges();
        return job;
    }

}

```

But I get the error `java.lang.NullPointerException: Cannot invoke "net.ravendb.client.http.NodeSelector.getPreferredNode()" because "this._nodeSelector" is null`

What am I doing wrong here ?

**Comment by ayende** [comment]
1/ a collection will be created the first time you save a document with that name.
From the Studio, there are two ways to do that.
You can save a document with the name `users/` - which will turn it into something like `users/0000000000000008883-A` in the `Users` collection.

Or you can include the following metadata:

```
    "@metadata": {
        "@collection": "Users"
    }
```

2/ From Java, the default is to take the pluralized name of your class. So saving a `User` instance will create a `User` document and create the `Users` collection if needed.

3/ You *cannot* create a singleton session. To start with, your code will return a session and dispose is store which is not valid.
Your session is meant to be transient, for a single request, while the `DocumentStore` is meant to be the singleton.

Using JPA / Hibernate terms, the `DocuemntStore` is the `SessionFactory` and the `DocumentSession` is the `Session`.

You code should be:

```
@Bean
IDocumentStore getDocumentStore() throws KeyStoreException, IOException, CertificateException, NoSuchAlgorithmException {
		KeyStore clientStore = KeyStore.getInstance("PKCS12");
		clientStore.load(new FileInputStream("C:\\certificate.with.password.pfx"), "D146AB87B64E3D2A23D3AD6606058A".toCharArray());
		DocumentStore store = new DocumentStore());
                store.setCertificate(clientStore);
		store.setDatabase("test");
		store.setUrls(new String[]{ "https://mydomain.ravendb.cloud" });
		return store.initialize();
	}

@RestController
public class TestController {
    @Autowired
    IDocumentStore store;

    @PostMapping("/test")
    public Job test(@RequestBody Job job) {
       try(IDocumentSession session = store.openSession()) {
            session.store(job);
            session.saveChanges();
            return job;
       }
    }

}
```

**Reply by rahul-raj** [reply]
@ayende 
Thanks for the info.
I tried your suggested approach  -> document store as singleton. But now I'm getting this error
`java.lang.IllegalStateException: The document store has already been disposed and cannot be used`

```
java.lang.IllegalStateException: The document store has already been disposed and cannot be used
	at net.ravendb.client.documents.DocumentStoreBase.ensureNotClosed(DocumentStoreBase.java:241) ~[ravendb-6.0.0.jar:na]
	at net.ravendb.client.documents.DocumentStore.openSession(DocumentStore.java:171) ~[ravendb-6.0.0.jar:na]
	at net.ravendb.client.documents.DocumentStore.openSession(DocumentStore.java:153) ~[ravendb-6.0.0.jar:na]
```

**Reply by ayende** [reply]
Did you keep it in the `try`? Remember, the `DocumentStore` is meant to be alive for as long as the app is.
Are you calling `close` (directly or indirectly)?

**Reply by rahul-raj** [reply]
This is my updated controller:
```
@RestController
public class TestController {
    @Autowired
    IDocumentStore store;

    @PostMapping("/test")
    public Job test(@RequestBody Job job) {
        try(IDocumentSession session = store.openSession()) {
            session.store(job);
            session.saveChanges();
            return job;
        }
    }

}
```

**Reply by ayende** [reply]
And the `store` init?

**Comment by rahul-raj** [comment]
I'm sorry, I got the problem and updated there.

Here's the updated init code:

```
@Bean
	IDocumentStore documentStore() throws KeyStoreException, IOException, CertificateException, NoSuchAlgorithmException {
		KeyStore clientStore = KeyStore.getInstance("PKCS12");
		clientStore.load(new FileInputStream("C:\\certificate.with.password.pfx"), "D146AB87B64E3D2A23D3AD6606058A".toCharArray());

		DocumentStore store = new DocumentStore();
        store.setCertificate(clientStore);
		store.setDatabase("test");
		store.setUrls(new String[]{ "https://myravendbspace.ravendb.cloud" });

		store.initialize();
		return  store;

	}
```

But now, I'm getting a different error
```
javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption
```

```
2024-06-09T14:03:26.373+05:30  INFO 23000 --- [demo] [ool-1-thread-10] n.ravendb.client.http.RequestExecutor    : Couldn't update topology from _updateTopologyTimer

java.util.concurrent.CompletionException: java.lang.IllegalStateException: Unable to configure ssl context: Get Key failed: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
Caused by: java.lang.IllegalStateException: Unable to configure ssl context: Get Key failed: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
	at net.ravendb.client.http.RequestExecutor.createClient(RequestExecutor.java:1776) ~[ravendb-6.0.0.jar:na]
Caused by: java.security.UnrecoverableKeyException: Get Key failed: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
```

**Reply by ayende** [reply]
It looks like your pfx file is problematic, or maybe with the wrong password?

**Reply by rahul-raj** [reply]
I used the password which was mentioned in the client certificate package.
Is it the right password ?
Also, I have imported the certificate (the one with password) and able to access RavenDB studio without issues.

**Reply by ayende** [reply]
What happens when you use the file without the password? 
What happens when you execute just this:

```
KeyStore clientStore = KeyStore.getInstance("PKCS12");
clientStore.load(new FileInputStream("C:\\certificate.with.password.pfx"), "D146AB87B64E3D2A23D3AD6606058A".toCharArray());
``` 

?

**Reply by rahul-raj** [reply]
I get the below error (**during application startup itself**) when I try to use the `certificate without password`:
```
Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.

```

I get the below error when I try to use `certificate with password` (**when I try to hit the REST API**):
```
javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
```

**Reply by ayende** [reply]
what happens when you use the certificate file that doesn't have a password?

```
clientStore.load(new FileInputStream("C:\\certificate.pfx"));
```

**Reply by rahul-raj** [reply]
![Screenshot 2024-06-09 125343](https://github.com/ravendb/ravendb/assets/517415/c831b520-d8a1-459b-9c0c-9359190c4857)
Getting compilation error there.
Can I use empty password instead with the previous method ? I'm using JDK 17

**Reply by ayende** [reply]
Are you available for a quick call? Would be easier than this back & forth.

**Reply by rahul-raj** [reply]
Yes, sure.

**Reply by ayende** [reply]
I'm here for the next 10 - 15 minutes

https://meet.google.com/rnb-gxoz-ymu

**Reply by rahul-raj** [reply]
@ayende 
Thanks a lot for the quick sync to expedite the issue.

The fix was to reference to PEM file (generated by double clicking on the PFX file) instead of the PFX file. I'm currently using Windows for development testing.

Code:
```
KeyStore clientStore =  CertificateUtils.createKeystore("C:\\certificate.pem");
```

Here is the link to documentation where I followed PFX based keystore setup which wasn't working:
https://ravendb.net/docs/article-page/6.0/java/client-api/setting-up-authentication-and-authorization

**Comment by ml054** [comment]
@rahul-raj I'd like to thank for sharing your case with us. I've added checks to DocumentStore initialization to prevent such issues in future. 
We will also update the documentation to be more clear in this regards. 

Happy coding with RavenDB. :)

Link: https://issues.hibernatingrhinos.com/issue/RavenDB-14344/Investigate-generation-of-certificates-with-multiple-aliases

**Reply by rahul-raj** [reply]
@ml054 
Thank you!
